### PR TITLE
duplicate entity validator

### PR DESCRIPTION
**Pull request by galkoren**

### CODE REVIEW

File: main/src/main/java/com/kenshoo/pl/entity/spi/helpers/DuplicateEntityValidator.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        this.uniqueKey = uniqueKey;
+    }
+
+    @Override
+    public SupportedChangeOperation getSupportedChangeOperation() {
+        return CREATE;
+    }
+
+    @Override
+    public void validate(Collection<? extends EntityChange<E>> commands, ChangeOperation op, ChangeContext ctx) {
```
Comment (UncleTalik) [inline main/src/main/java/com/kenshoo/pl/entity/spi/helpers/DuplicateEntityValidator.java:34]: LGTM, but to small suggestions:
1. Essentially this validation is checking two things: a) duplicates within the provided commands collections. b) duplicates in the DB. SO: maybe extract these to two separate methods and then the validate will look something like this:
``` Java
Map<Identifier<E>, ? extends EntityChange<E>> commandsByIds = markDuplicatesInCollectionWithError(commands);
markDuplicatedToDbWithError(commandsByIds);
```   
2. I would suggest to add a test here. mainly testing the following scenarios/use cases:
noValidationWhenNoDuplicateExists
markDuplicatesWithinProvidedCommandsWtihError 
markDuplicatesToDbWtihError 
markDuplicatesToDbAndInProvidedCollectionWtihErrors

File: integration-tests/src/test/java/com/kenshoo/pl/entity/spi/helpers/UniquenessValidatorTest.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        ALL_TABLES.forEach(table -> dslContext.deleteFrom(table).execute());
+    }
+
+    @AfterClass
+    public static void dropTables() {
+        ALL_TABLES.forEach(table -> staticDSLContext.dropTableIfExists(table).execute());
+    }
+
+    @Test
+    public void test_dont_fail_if_key_is_unique() {
```
[the remark this answers is not in the document]
Comment (galkoren) [reply integration-tests/src/test/java/com/kenshoo/pl/entity/spi/helpers/UniquenessValidatorTest.java]: Tal!! I accept this suggestion

File: main/src/main/java/com/kenshoo/pl/entity/spi/helpers/DuplicateEntityValidator.java (hunk 1 above)
Comment (galkoren) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/DuplicateEntityValidator.java]: Hi man.
1. Will do
1. I already have such tests...

Comment (UncleTalik) [reply main/src/main/java/com/kenshoo/pl/entity/spi/helpers/DuplicateEntityValidator.java]: Yeah, I see the tests now. I think I wrote the comment before there was a test class :)

File: integration-tests/src/test/java/com/kenshoo/pl/entity/spi/helpers/UniquenessValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+                        .with(ParentEntity.ID_IN_TARGET, 3333)
+        );
+
+        CreateResult<ParentEntity, Identifier<ParentEntity>> results = parentPersistence.create(commands, parentFlow(validator).build());
+
+        assertTrue(results.hasErrors());
+    }
+
+    @Test
+    public void test_dont_fail_command_condition_is_unmatched() {
```
[the remark this answers is not in the document]
Comment (galkoren) [reply integration-tests/src/test/java/com/kenshoo/pl/entity/spi/helpers/UniquenessValidatorTest.java]: It is useful for the KS.
The KS never delete entities but rather only mark them as "Deleted".
So we need this:

```java
.Builder<>(entitiesFetcher, uniqueness)
                .setCondition(PLCondition.not(MyEntity.STATUS_IN_TRAGET.eq(Deleted)))
                .build();
```

File: main/src/main/java/com/kenshoo/pl/entity/spi/helpers/UniquenessValidator.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    private final String errorCode;
+    private final EntitiesFetcher fetcher;
+    private final UniqueKey<E> uniqueKey;
+    private final PLCondition condition;
+
+    private UniquenessValidator(EntitiesFetcher fetcher, UniqueKey<E> uniqueKey, PLCondition condition, String errorCode) {
+        this.errorCode = errorCode;
+        this.fetcher = requireNonNull(fetcher, "entity fetcher must be provided");
+        this.uniqueKey = requireNonNull(uniqueKey, "unique key must be provided");
+        this.condition = requireNonNull(condition, "condition must be provided");
```
Comment (effiban) [inline main/src/main/java/com/kenshoo/pl/entity/spi/helpers/UniquenessValidator.java:42]: :+1:

File: integration-tests/src/test/java/com/kenshoo/pl/entity/spi/helpers/UniquenessValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+    private List<CreateEntityCommand<ParentEntity>> failedCommands(CreateResult<ParentEntity, Identifier<ParentEntity>> results) {
+        return results.getChangeResults().stream().filter(not(r -> r.isSuccess())).map(r -> r.getCommand()).collect(toList());
+    }
+
+    private ChangeFlowConfig.Builder<ParentEntity> parentFlow(ChangesValidator<ParentEntity>... validators) {
+        return ChangeFlowConfigBuilderFactory
+                .newInstance(plContext, ParentEntity.INSTANCE)
+                .withValidators(Seq.of(validators).toList())
+                ;
```
[the remark this answers is not in the document]
Comment (effiban) [reply integration-tests/src/test/java/com/kenshoo/pl/entity/spi/helpers/UniquenessValidatorTest.java:202]: @galkoren - this is still TBD

**Pull request merged**